### ISSUE TITLE
Fog or frost potentially not working

### ISSUE THREAD

**Issue by wlbragg** [c172p-team/c172p#910]
Could someone else verify this. I can't seem to get any fog or frost to generate. Maybe a hint as to a good METAR might help?

**Comment by wlbragg** [comment]
Here is where I am at so far. I added this system to the Cub and that didn't appear to work. . I wanted to make sure it was the Cub implementation and not something else. So I tested the C172P system and I also couldn't generate fog or ice. I went bake to the Cub and was working on stuff while flying for over an hour with the window open and a temp of around -4 C. Eventually I did develop frost, so I guess that is working. After I shut the window and applied the cabin heat and opened the vent , the frost dissipated quickly. Still no fog even though I tried several METAR combinations. I still think it's a good idea if someone else can confirm the fog system is working correctly before we close this issue. Shouldn't we be able to set up the environment to stimulate the  creation of fog?

**Comment by wlbragg** [comment]
Well this is strange, suddenly it appears that I am now getting fog and frost, and enough of both I don't think the cabin heat can keep up (this is in the Cub). So I guess this is working. I have to quit for the night, I would still really like to test the c172p tomorrow and make sure it is working before closing. In the meantime, any tips on setting up the environment to facilitate testing this system would be really helpful.

**Comment by onox** [comment]
> I went back to the Cub and was working on other stuff while flying for over an hour with the window open and a temp of around -4 C

If you're flying with open windows, it's hard to get warm air in the cabin. For fog you need cold windshield + warm air IIRC.

**Comment by onox** [comment]
You need to look at `/fdm/jsbsim/heat/surface-dewpoint-diff` (which is equal to `/fdm/jsbsim/heat/cabin-dewpoint-degc` - `/fdm/jsbsim/heat/surface-temp-degc`). It needs to be positive for fog to develop.

**Comment by onox** [comment]
Relative humidity is also important. Go to a place where it's humid, not necessarily cold. Add some pax to add additional humidity.

**Comment by wlbragg** [comment]
There is one area I am concerned with. It relates to the warning to turn the heat down to reduce the fogging. I get the principle that adding "moist" heat into the cockpit increases the temperature diff between the inside and the outside. The only issue I have is in the case of a vent and heat being blown out the vent close to or on the surface of the window. That is how we remove fogging in a vehicle. How did you handle this? Opening the heat vent should reduce window fogging, not increase it?

**Comment by wlbragg** [comment]
>when the temperature of the glass drops below the dew point of the inside air, visible moisture forms on the glass. The combination of a high level of moisture in the air and a cool glass surface triggers the condensation process.

The heating of the glass reduces the ability of condensation to form. For condensation you need moist air and cold surfaces. So when introducing hot air into the cabin there would be a tug-a-war between temperature gradient between air and glass (fogging) and the rate at which the glass heated to match the interior air temp (remove fogging).

**Comment by onox** [comment]
> The only issue I have is in the case of a vent and heat being blown out the vent close to or on the surface of the window. That is how we remove fogging in a vehicle. How did you handle this? Opening the heat vent should reduce window fogging, not increase it?

I did not model the adjustable defroster outlets because they don't exist in the 3D model. Do we need to paint one of the R/G/B layers in the ALS glass mask? (so that you see the fog disappearing where the defroster outlets are)

**Comment by onox** [comment]
See figure 7-8 on page 7-31 in the POH.

**Comment by wlbragg** [comment]
>Do we need to paint one of the R/G/B layers in the ALS glass mask

Yes that was always a goal.

Something else, the property /fdm/jsbsim/heat/cabin-air-temp-degc, we initialize that property at 0.0 when in reality it appears it should initially be set at whatever the outside temp is at startup?

**Comment by wlbragg** [comment]
If that is the interior cabin temperature, at startup, it needs to start at the interior temperature equalized with the outside temperature, not at 0C. Otherwise we are constantly fighting an interior temp that starts at 0C.

**Comment by wlbragg** [comment]
Back to my original concern...

>I did not model the adjustable defroster outlets because they don't exist in the 3D model.

Besides that, I guess you also have the interior glass temperature increase as the cabin temp increases, but at a slow, defined rate. That leads to another equalization issue. Don't we need to also have the glass temperature start out equal to the outside air temp?

You have two integrators

~~~xml
<integrator name="heat/cabin-air-temp-degc">
            <input>heat/cabin-air-transfer-total</input>

            <!-- This is the constant that actually controls how fast
                 (or slow) the cabin temperature changes.
            -->
            <c1>0.030</c1>
        </integrator>
~~~

and

~~~xml
        <integrator name="heat/surface-temp-degc">
            <input>heat/surface-cabin-heat-transfer</input>

            <!-- This is the constant that actually controls how fast
                 (or slow) the surface temperature changes.
            -->
            <c1>0.020</c1>
        </integrator>
~~~

Don't both those properties have to start out at the same temp as /environment/temperature-degc?

I don't totally understand the integrator function, (I'm looking it up now), but it looks like your starting out with a cabin air temperature based on a combination of all the influencing heat sources (because that is how I see you creating the property "environment/temperature-degc"), or did I miss something? I am saying that it should start out as an equalized value that matches the outside air temp, not based on engine heat, body heat, etc., those values would influence it from that point on, after it was initialized at the same temp as the outside air temp upon startup.

**Comment by wlbragg** [comment]
When starting out in the c172p, this is the starting values
cabin temp = heat/cabin-air-temp-degc = 1.94C
outside temp = environment/temperature-degc = 17.25C

Am I not understanding the use of the property heat/cabin-air-temp-degc?

**Comment by onox** [comment]
> If that is the interior cabin temperature, at startup, it needs to start at the interior temperature equalized with the outside temperature, not at 0C. Otherwise we are constantly fighting an interior temp that starts at 0C.

`heat/cabin-air-temp-degc` is output by an integrator block. There isn't a way to initialize it as far as I know. We've talked about this thing. You should find it being mentioned in old issues.

**Comment by wlbragg** [comment]
But that leave this whole thing incorrect doesn't it? I don't see how it can work starting out with a value in the cabin and a glass temp of 0C or 32F no matter the environmental temp.

**Comment by onox** [comment]
@wlbragg See last paragraph of https://github.com/c172p-team/c172p-detailed/pull/576#issuecomment-161404443

**Comment by onox** [comment]
And last paragraph of https://github.com/c172p-team/c172p-detailed/pull/576#issuecomment-162318858

**Comment by wlbragg** [comment]
Aaah!

I guess my memory is failing me :cry: 

I got distractions happening around me so i have to stop for a short while. I want to look into this environmental temp issue and see what your referring to. I think we have to have it to make this work right.

**Comment by onox** [comment]
This discussion is giving me a deja vu 😜

**Comment by wlbragg** [comment]
Sorry @onox, after reading through the prior issue it is crystal clear now what's going on. I wish I had remembered this right at the start, it would have saved me some time. Do you have any idea approximately how long it takes to equalize a 30-40 degree difference and start working correctly?

I saw where you already commented on my concern above in the prior issue comments, ie:
> If you increase the temperature (using cabin heat) then cabin temperature + dew point goes up. So you should get some fog then. After a while, the surface temperature catches up (it changes more slowly) and becomes lower than cabin dew point, so fog disappears again.

So that answers my concern. I still think it would be cool to have the vent mask.

Also if the environment temp is out of whack, then even starting at 0 isn't going to help, it will still stay mucked up because the temps are not accurate. So with that in mind, I think we should still sum the integrator output with the initial environment temperature. I am going to do the Cub that way and test it thoroughly and see if I end up at the same conclusion you did.

**Comment by wlbragg** [comment]
Wow, what it is, is the weather has to settle. It is taking consistently 1-2 minutes to settle into it's stable value.

**Comment by wlbragg** [comment]
Yep, it's an absolute can of worms because the weather has to settle, :angry:

**Comment by onox** [comment]
Yeah, it is unfortunate. But adding the defroster outlets would be cool though. If you can get the ALS glass mask to work, I'll add support for it in the code.

**Comment by wlbragg** [comment]
@onox can you use the value from a property before that property is created chronologically speaking in the file order?

We have early on in the file that uses heat/cabin-air-temp-degc
 ~~~xml
<summer name="heat/cabin-heat-temp-diff">
            <input>/environment/temperature-degc</input>
            <input>-heat/cabin-air-temp-degc</input>
            <input>heat/heat-exhaust-gas-temp-degc</input>
        </summer>
~~~

but the heat/cabin-air-temp-degc as far as I can tell is created later in the file
~~~xml
<integrator name="heat/cabin-air-temp-degc">
            <input>heat/cabin-air-transfer-total</input>
            
            <!-- This is the constant that actually controls how fast
                 (or slow) the cabin temperature changes.
            -->
            <c1>0.030</c1>
        </integrator>
~~~

Is it simply NULL until it is created and then it contains a value?

**Comment by onox** [comment]
> can you use the value from a property before that property is created chronologically speaking in the file order?

Yes, the summer calls `SGPropertyNode::getDoubleValue()` and that function call `SGRawValue<double>::DefaultValue()` (which returns 0.0) if it doesn't know the type.

**Comment by onox** [comment]
@wlbragg I checked JSBSim's source code again and the integrators are always initialized to 0. Could you ask the JSBSim devs on the mailing list to add something like:

```xml
<init>/environment/temperature-degc</init>
```

**Comment by wlbragg** [comment]
Yes, but let me make sure I understand what we're asking for.
~~~
<integrator name="heat/cabin-air-temp-degc">
            <input>heat/cabin-air-transfer-total</input>
            
            <!-- This is the constant that actually controls how fast
                 (or slow) the cabin temperature changes.
            -->
            <c1>0.030</c1>
        </integrator>
~~~
In this example the integrator "heat/cabin-air-temp-degc" is initialized to 0 and takes x time to adjust to any value. We want to be able to initialize it immediately to a value, in this case we want to use a property to initialize it. So were asking for a tag <init></init> that you can set with a value or with a property to be a part of the integrator function. Correct?

**Comment by onox** [comment]
Yes. But we could also fix it by summing the output to some property that we set just once. But then we need to take care of it manually when FG's is reset etc. So maybe you could phrase it as a question: what do they recommend?

**Comment by wlbragg** [comment]
@onox 
We received a reply a week or two ago, sorry I am just now forwarding this.

>Re: [Flightgear-devel] Possible enhancement to JSBsim integrators              
From: Bertrand Coconnier <bcoconni@gm...> - 2018-02-24 18:28:44                                        
--
Hi,  During its initialization, <integrator> uses whatever value is in the <input> property has its initial value. Have you tried to initialize the property heat/cabin-air-transfer-total during the initialization process ? In order to work, this must happen before JSBSim loads.  Bertrand.

**Comment by onox** [comment]
The problem is AW may initialize for a long time after everything else (including the FDM) has been loaded.

**Comment by wlbragg** [comment]
James proposed this on dev list for a different issue but may be of help here for initializing the fog/ice value. 
I was also wondering about another potential issue on an aircraft parked outside all night, dew? This would potentially allow for a function for that as well.
